pkg/cloud/prepare/aws.go

```diff
@@ -54,7 +54,11 @@ func AWS(clusterInfo *cluster.Info, ports *cloud.Ports, config *aws.Config, useL
 				}
 			}
 
-			return cloud.PrepareForSubmariner(input, status)
+			if len(input.InternalPorts) > 0 {
+				return cloud.PrepareForSubmariner(input, status)
+			}
```

Comment:
It seems like this validation should be done in `cloud.PrepareForSubmariner` otherwise this code is making assumptions about what `cloud.PrepareForSubmariner` does internally, ie calling it in a 'white box' manner rather than 'black box'.  We'd have a latent bug if the behavior of `cloud.PrepareForSubmariner` ever changed such that it did more than open internal ports.

Reply:
Actually right now (and for a long while) it's only opening internal ports so it makes sense to do so here and rename that function in cloud prepare itself if we're worried about that.

Reply:
We know that, but `subctl` isn’t supposed to. The description of `PrepareForSubmariner` is very generic, and just says it prepares a cloud for Submariner. The AWS version already handles this correctly, so there is precedent for handling empty `InternalPorts` slices in the `PrepareForSubmariner` function — and I’d argue that *that* makes sense, there’s no documented requirement that `PrepareForSubmariner` be called only with non-empty `InternalPorts`.

Reply:
I propose we go with https://github.com/submariner-io/cloud-prepare/pull/571 and leave this as is.

In the same manner, we only try to deploy gateways when gateways have been requested.

The whole `PrepareForSubmariner` made sense initially when cloud prepare was just incepted, but due to limitations it has been split out to internal & external preparation, and realistically ever since it started it just opens internal ports.
Using a simpler API makes it clear and avoids any confusion that might arise, while keeping it simple.

Reply:
It's fine to rename the API but I still think the implementation should handle the trivial case of an empty port list to avoid all users having to do it. As @skitt mentioned, the AWS implementation does check but others don't - we should be consistent.

Reply:
I think while it needs to be fixed in respective implementations as well, we can have this check too